File: src/main/java/me/bristermitten/plumber/scheduling/Task.java
package me.bristermitten.plumber.scheduling;

import com.google.inject.assistedinject.Assisted;
import com.google.inject.assistedinject.AssistedInject;
import me.bristermitten.plumber.PlumberPlugin;
import me.bristermitten.plumber.object.Resettable;
import me.bristermitten.plumber.scheduling.timings.Time;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

public class Task implements Runnable, Resettable {

    private final long delay;
    private final long period;

    private final Runnable delegate;

    private final PlumberPlugin plugin;
    private BukkitTask task;

    @AssistedInject
    public Task(PlumberPlugin plugin, @Assisted("delay") long delay, @Assisted("period") long period,
                @Assisted Runnable delegate) {
        this.plugin = plugin;
        this.delay = delay;
        this.period = period;
        this.delegate = delegate;
    }

    @AssistedInject
    public Task(PlumberPlugin plugin, @Assisted("delay") Time delay, @Assisted("period") Time period,
                @Assisted Runnable delegate) {
        this.plugin = plugin;
        this.delay = delay.toTicks();
        this.period = period.toTicks();
        this.delegate = delegate;
    }

    public final void start() {
        BukkitRunnable runnable = new BukkitRunnable() {
            @Override
            public void run() {
                Task.this.run();
            }
        };
        if (delay == Time.Companion.getNONE_TICKS() && period == Time.Companion.getNONE_TICKS()) {
            task = runnable.runTaskAsynchronously(plugin);
            return;
        }
        if (delay == Time.Companion.getNONE_TICKS()) {
            task = runnable.runTaskTimerAsynchronously(plugin, 0, period);
            return;
        }
        if (period == Time.Companion.getNONE_TICKS()) {
            task = runnable.runTaskLaterAsynchronously(plugin, delay);
            return;
        }
        task = runnable.runTaskTimerAsynchronously(plugin, delay, period);
    }

    public final void stop() {
    }

    @Override
    public void run() {
        delegate.run();
    }

    @Override
    public void reset() {
        if (task != null && !task.isCancelled()) {
            task.cancel();
        }
    }
}

File: src/main/java/me/bristermitten/plumber/scheduling/Task.kt
package me.bristermitten.plumber.scheduling

import com.google.inject.assistedinject.Assisted
import com.google.inject.assistedinject.AssistedInject
import me.bristermitten.plumber.PlumberPlugin
import me.bristermitten.plumber.struct.Resettable
import me.bristermitten.plumber.scheduling.timings.Time
import org.bukkit.scheduler.BukkitTask

/**
 * A Task wraps Bukkit's Scheduler system into a single object.
 * Currently, all tasks are ran asynchronously, but this will be improved in the future.
 *
 * Generally, all Tasks are created with [TaskFactory]
 *
 */
class Task : Runnable, Resettable {

    private val delay: Long
    private val period: Long
    private val run: Runnable

    private val plugin: PlumberPlugin

    private lateinit var task: BukkitTask

    @AssistedInject
    constructor(
            plugin: PlumberPlugin,
            @Assisted("delay") delay: Long,
            @Assisted("period") period: Long,
            @Assisted delegate: Runnable
    ) {
        this.plugin = plugin
        this.delay = delay
        this.period = period
        this.run = delegate
    }

    @AssistedInject
    constructor(
            plugin: PlumberPlugin,
            @Assisted("delay") delay: Time,
            @Assisted("period") period: Time,
            @Assisted delegate: Runnable
    ) {
        this.plugin = plugin
        this.delay = delay.toTicks()
        this.period = period.toTicks()
        this.run = delegate
    }

    fun start() {
        val scheduler = plugin.server.scheduler
        when {
            delay == Time.NONE_TICKS && period == Time.NONE_TICKS -> {
                task = scheduler.runTaskAsynchronously(plugin, this)
                return
            }
            delay == Time.NONE_TICKS -> {
                task = scheduler.runTaskTimerAsynchronously(plugin, this, 0, period)
                return
            }
            period == Time.NONE_TICKS -> {
                task = scheduler.runTaskLaterAsynchronously(plugin, this, delay)
                return
            }
            else -> task = scheduler.runTaskTimerAsynchronously(plugin, this, delay, period)
        }
    }

    fun stop() {
        task.cancel()
    }

    override fun run() {
        run.run()
    }

    override fun reset() {
        if (!task.isCancelled) {
            task.cancel()
        }
    }
}
